services/snapshotter/client.go

```diff
@@ -22,6 +26,138 @@ func NewClient(host string) *Client {
 	return &Client{host: host}
 }
 
+// takes a request object, writes a Base64 encoding to the tcp connection, and then sends the request to the snapshotter service.
+// returns a mapping of the uploaded metadata shardID's to actual shardID's on the destination system.
+func (c *Client) UpdateMeta(req *Request, upStream io.Reader) (map[uint64]uint64, error) {
+	var err error
+
+	var b bytes.Buffer
+
+	// Connect to snapshotter service.
+	conn, err := tcp.Dial("tcp", c.host, MuxHeader)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	if _, err := conn.Write([]byte{byte(req.Type)}); err != nil {
+		return nil, err
+	}
+
+	if err := json.NewEncoder(conn).Encode(req); err != nil {
+		return nil, fmt.Errorf("encode snapshot request: %s", err)
+	}
+
+	if n, err := io.Copy(conn, upStream); (err != nil && err != io.EOF) || n != req.UploadSize {
+		return nil, fmt.Errorf("error uploading file: err=%v, n=%d, uploadSize: %d", err, n, req.UploadSize)
+	}
+
+	if n, err := b.ReadFrom(conn); err != nil || n == 0 {
+		return nil, fmt.Errorf("updating metadata on influxd service failed: err=%v, n=%d", err, n)
+	}
+
+	resp := b.Bytes()
+
+	if err != nil {
+		return nil, err
+	}
+
+	header, npairs, err := c.decodeUintPair(resp[:16])
+	if err != nil {
+		return nil, err
+	}
+
+	if npairs == 0 {
+		return nil, fmt.Errorf("DB metadata not changed. database may already exist")
+	}
+
+	pairs := resp[16:]
+
+	if header != BackupMagicHeader {
+		return nil, fmt.Errorf("Response did not contain the proper header tag.")
+	}
+
+	if (len(pairs)/8)%2 != 0 {
+		return nil, fmt.Errorf("expected an even number of integer pairs in update meta repsonse")
+	}
+
+	shardIDMap := make(map[uint64]uint64)
+	for i := 0; i < int(npairs); i++ {
+		offset := i * 16
+		k, v, err := c.decodeUintPair(pairs[offset : offset+16])
+		if err != nil {
+			return nil, err
+		}
+		shardIDMap[k] = v
+	}
+
+	return shardIDMap, nil
+}
+
+func (c *Client) decodeUintPair(bits []byte) (uint64, uint64, error) {
+	if len(bits) != 16 {
+		return 0, 0, fmt.Errorf("array must have exactly 16 bytes")
+	}
+	v1 := binary.BigEndian.Uint64(bits[:8])
+	v2 := binary.BigEndian.Uint64(bits[8:16])
+	return v1, v2, nil
+}
+
+func (c *Client) UploadShard(shardID, newShardID uint64, destinationDatabase, restoreRetention string, tr *tar.Reader) error {
+
+	conn, err := tcp.Dial("tcp", c.host, MuxHeader)
+	defer conn.Close()
+	if err != nil {
+		return err
+	}
+
+	if _, err := conn.Write([]byte{byte(RequestShardUpdate)}); err != nil {
+		return err
+	}
+
+	var shardBytes [8]byte
```

Comment:
Minor nit: make this `[9]byte` and put `RequestShardUpdate` in the first byte, then encode the ID into `shardBytes[1:9]`. You're currently making two small writes of known length and content back-to-back, it would be better to combine them.